content/blog/2020-12-20-bril-algebraic-data-types/index.md

```diff
@@ -0,0 +1,538 @@
+The JSON format of the `pack` instruction is a follows:
+
+```json
+{
+    "op": "pack",
+    "args": ["a1", ..., "an"],
+    "dest": "d",
+    "type": {"product": [t1, ..., tn]}
+}
+```
+
+The textual format of the `pack` instruction is as follows:
+
+```
+d: product<t1, ..., tn> = pack a1 ... an
+```
+
+#### `unpack`
+
+Conversely, the `unpack` instruction is used to extract an individual
+value from a product-typed value. It takes two arguments: a variable
+(which must be a tuple), and the (zero-indexed) index of the value
+from the tuple to extract.
+
+Given that the input tuple has type `product<t1, ..., tn>` and the
+index is `i`, the destination must have type `ti`.
+
+The JSON format of the `unpack` instruction is as follows:
+
+```json
+{
+    "op": "unpack",
+    "args": ["a", "i"],
+    "dest": "d",
+    "type": "t"
+}
+```
+
+The text format of the `unpack` instruction is as follows:
+
+```
+d: t = unpack a i
+```
+
+#### Example
+
+Here is a sample Bril program that uses tuples:
+
+```
+@main() {
+  one: int = const 1;
+  two: int = const 2;
+  tuple: product<int, int> = pack one two;
+  one_again = unpack tuple 0;
+}
+```
+
+### Sum types
+
+#### `construct`
+
+The `construct` instruction is used to construct sum types. It takes
+as its sole argument the value to use. The semantics of the
+instruction dictate that constructor to use is determined
+automatically by examining the runtime type of the argument.
+
+The destination must have a sum type and the type of the argument must
+correspond to one of the constructors of that sum type.
+
+The JSON format of the `construct` instruction is as follows:
+
+```json
+{
+    "op": "construct",
+    "args": ["a"],
+    "dest": "d",
+    "type": {"sum": [t1, ..., tn]}
+}
+```
+
+The text format of the `construct` instruction is as follows:
+
+```
+d: sum<t1, ..., tn> = construct a
+```
+
+In theory, it could have been possible to omit `construct` entirely if
+there were a sum type subtyping rule in place, but such a rule would
+probably have made things more, not less, confusing.
+
+#### `destruct`
+
+The `destruct` instruction is used to execute different code based on
+the constructor used to create a sum-typed value. It takes a sum-typed
+value as an argument and a number of labels, one for each constructor
+in the sum type.
+
+The destination of the instruction is a variable name with a sum
+type. The semantics of the instruction dictate that the control flow
+jumps to the correct branch based on the argument and that the
+destination value gets assigned the value within the constructor.
+
+The destination will in fact not have the same sum type as the
+argument; it will have the type carried by one of the constructors,
+which is different for each label. This is confusing because the
+destination is annotated with a type that it cannot have, but I chose
+this approach because the Bril interpreter needs to know the available
+constructors in order to select the correct label to jump to.
```

Comment:
Yeah, this is pretty weird! I wonder what a way around this would have been. One option would be to do the "unsafe projection" thing. So you'd say "get me the `int` from this `sum<int, float>`" and it would give you an int or just crash. And there would be a separate instruction to say "which variant are you?" that would presumably return an integer.